src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,209 @@
+        totpStorage.insertUsedCode(newCode);
+
+        // Check if the code is valid:
+        if (!checkCode(matchingDevice, code)) {
+            throw new InvalidTotpException();
+        }
+
+        // Check if the code is unused:
+        TOTPUsedCode[] usedCodes = totpStorage.getUsedCodes(userId);
+        for (TOTPUsedCode usedCode : usedCodes) {
+            if (usedCode.code.equals(code) && usedCode.isValidCode) {
+                throw new InvalidTotpException();
+            }
+        }
+
+        totpStorage.markDeviceAsVerified(userId, deviceName);
+        return new VerifyDeviceResponse(false);
+    }
+
+    public static void verifyCode(Main main, String userId, String code, boolean allowUnverifiedDevices)
+            throws StorageQueryException, TotpNotEnabledException, InvalidTotpException,
+            LimitReachedException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // FIXME: Every unused_code should be linked to the device it was used for.
+        // Otherwise, if a user has multiple devices, and they use the same code for
+        // both,
+        // then the code will be considered as used for both devices. This could cause
+        // UX
+        // issues.
+        // If we do this, then it also means that we need to assign a device ID to each
+        // device (OR use
+        // (userId, deviceName) as the ID)
+
+        // Check if the code has been successfully used by the user (for any device):
+        TOTPUsedCode[] usedCodes = totpStorage.getUsedCodes(userId);
+        for (TOTPUsedCode usedCode : usedCodes) {
+            if (usedCode.code.equals(code) && usedCode.isValidCode) {
+                throw new InvalidTotpException();
+            }
+        }
+
+        // Try different devices until we find one that works:
+        boolean isValid = false;
+        for (TOTPDevice device : devices) {
+            // Check if the code is valid for this device:
+            if (checkCode(device, code)) {
+                isValid = true;
+                break;
+            }
+        }
+
+        // Insert the code into the list of used codes:
+        TOTPUsedCode newCode = new TOTPUsedCode(userId, code, isValid, System.currentTimeMillis() + 1000 * 60 * 5);
+        totpStorage.insertUsedCode(newCode);
+
+        if (isValid) {
+            return;
+        }
+
+        // Check if last 5 codes are all invalid:
+        int WINDOW_SIZE = 5;
+        int invalidCodes = 0;
+        for (int i = usedCodes.length - 1; i >= 0 && i >= usedCodes.length - WINDOW_SIZE; i--) {
+            if (!usedCodes[i].isValidCode) {
+                invalidCodes++;
+            }
+        }
```

Comment:
Make sure that this is very well tested

Reply:
Added 2 tests in `createDeviceAndVerifyCode`:
- Trigger rate limit, assert error types, and then pass a valid code to ensure that rate limit error stops.
- Trigger rate limit, assert error types, and then run expired token cron to ensure that rate limit error stops.